### PR TITLE
Add segment_max operator.

### PR DESCRIPTION
**Pull request by alexdavies**
Segment_max is an important operation for GraphNets and is an open feature request at https://github.com/google/jax/issues/2255

### CODE REVIEW

**Review by jekbradbury — APPROVED** [review]
This looks good, sorry for the delay! I think existing tests for ops/scatter.py are in [lax_numpy_indexing_test](https://github.com/google/jax/blob/master/tests/lax_numpy_indexing_test.py#L960), so that's why there isn't a scatter_test.py yet. We probably don't want tests in two places, so you can either move other scatter.py related tests to scatter_test.py or add your test to lax_numpy_indexing_test.py

**Comment by alexdavies** [comment]
Thanks! Resolved by putting the test in lax_numpy_indexing_test.

File: jax/ops/scatter.py [hunk 1]
```diff
@@ -308,3 +309,37 @@ def segment_sum(data, segment_ids, num_segments=None):
...
+    the max of the associated segments.
+  """
+  if num_segments is None:
+    num_segments = jnp.max(jnp.mod(segment_ids, data.shape[0])) + 1
+  num_segments = int(num_segments)
+
+  full_zeros = jnp.zeros((num_segments,) + data.shape[1:], dtype=data.dtype)
+  full_ones = jnp.ones((num_segments,) + data.shape[1:], dtype=data.dtype)
+
+  segment_ids = jnp.mod(segment_ids, num_segments)
```
Comment (hawkinsp) [inline jax/ops/scatter.py:340]: I'm wondering whether this is wise. Divisions can be expensive; I realize you are probably just copying `segment_sum` but I'm thinking users should not come to expect this "wrapping" out of domain behavior.

(FWIW, the semantics of `index_update` is that updates with indices out of range are dropped.)

File: jax/ops/scatter.py [hunk 2]
```diff
@@ -308,3 +309,37 @@ def segment_sum(data, segment_ids, num_segments=None):
...
+
+  valid_updates = index_update(full_zeros, segment_ids, jnp.ones_like(data))
```
Comment (hawkinsp) [inline jax/ops/scatter.py:342]: Given this is a mask, perhaps do the scatter with a `bool` type?

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Sorry about the slow review.

File: jax/ops/scatter.py (hunk 1 above)
Comment (alexdavies) [reply jax/ops/scatter.py:340]: I'd be very happy to drop that and move to a more straightforward implementation - I did just include this for consistency with the existing segment_sum implementation.

File: jax/ops/scatter.py (hunk 2 above)
Comment (alexdavies) [reply jax/ops/scatter.py:342]: Good idea

File: jax/ops/scatter.py [hunk 3]
```diff
@@ -308,3 +309,35 @@ def segment_sum(data, segment_ids, num_segments=None):
   out = jnp.zeros((num_segments,) + data.shape[1:], dtype=data.dtype)
   segment_ids = jnp.mod(segment_ids, num_segments)
   return index_add(out, segment_ids, data)
+
+
+def segment_max(data, segment_ids, num_segments=None):
+  """Computes the max within segments of an array.
+
+  Similar to TensorFlow's segment_max:
+  https://www.tensorflow.org/api_docs/python/tf/math/segment_max
```
Comment (hawkinsp) [inline jax/ops/scatter.py:318]: You may need a blank line before `Args`.

I wanted to check the documentation rendering and then I realized this PR also needs to add `segment_max` to the docs. (`jax.ops.rst`)

File: tests/lax_numpy_indexing_test.py [hunk 4]
```diff
@@ -982,6 +982,42 @@ def testSegmentSum(self):
     expected = onp.array([13, 2, 7, 4])
     self.assertAllClose(ans, expected, check_dtypes=False)
 
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name": "_val={}_idx={}".format(
+          jtu.format_shape_dtype_string(data_shape, val_dtype),
+          jtu.format_shape_dtype_string(data_shape, idx_dtype)),
+       "val_dtype": val_dtype, "idx_dtype": idx_dtype,
+       "rng_idx_factory": rng_idx_factory,
```
Comment (hawkinsp) [inline tests/lax_numpy_indexing_test.py:990]: There's actually no need to make these test parameters if there's only one of them. (The reason many test cases put these as parameters is historical.)

File: tests/lax_numpy_indexing_test.py [hunk 5]
```diff
@@ -982,6 +982,42 @@ def testSegmentSum(self):
...
+      }
+      for idx_dtype in int_dtypes
+      for val_dtype in default_dtypes
+      for data_shape in [(3,), (3, 4), (3, 4, 4)]
+      for rng_idx_factory in [partial(jtu.rand_int, high=data_shape[0])]
+      for rng_factory in [jtu.rand_default]))
+  def testSegmentMax(self, val_dtype, idx_dtype, rng_idx_factory, rng_factory,
+                     data_shape):
+    def args_maker():
+      rng = rng_factory(onp.random.RandomState(0))
```
Comment (hawkinsp) [inline tests/lax_numpy_indexing_test.py:1002]: You should instantiate these outside `args_maker`.

And I think you could instantiate both with `self.rng()`, which is a random state deterministically chose based on the full test case name (including any parameters, etc.)

**Review by hawkinsp — CHANGES_REQUESTED** [review]
Sorry, I didn't see you had pushed changes until you rerequested a review.

**Comment by Joshuaalbert** [comment]
This is still an important feature for softmax reducer in graph networks.

**Comment by hawkinsp** [comment]
Looks like `jax.ops.segment_max` was added by another PR. Thanks for helping make this happen!

**Comment by hawkinsp** [comment]
(https://github.com/google/jax/pull/6384 was the PR that added it)
